src/LinAlg/hiopLinSolverSparseCUSOLVER.cpp

```diff
@@ -0,0 +1,941 @@
+                                     /* A is original matrix */
+                                     nnz_,
+                                     descr_A_,
+                                     da_,
+                                     dia_,
+                                     dja_,
+                                     devr_, /* right hand side */
+                                     devx_,   /* left hand side */
+                                     &ite_refine_succ_,
+                                     &r_nrminf_,
+                                     info_M_,
+                                     d_work_);
+    //copy the solutuion back: dx = devx_
+    if(sp_status_ == 0){
+      checkCudaErrors(cudaMemcpy(dx, devx_, sizeof(double) * n_, cudaMemcpyDeviceToHost));
+    }
+    nlp_->runStats.linsolv.tmTriuSolves.stop();
+
+    delete rhs; rhs = nullptr;
+    return 1;
+  }
+
+  //
+  // The Solver for Nonsymmetric KKT System
+  //
+
+  hiopLinSolverNonSymSparseCUSOLVER::hiopLinSolverNonSymSparseCUSOLVER(const int& n, const int& nnz, hiopNlpFormulation* nlp)
+    : hiopLinSolverNonSymSparse(n, nnz, nlp),
+      kRowPtr_{nullptr},
+      jCol_{nullptr},
+      kVal_{nullptr},
+      index_covert_CSR2Triplet_{nullptr},
+      index_covert_extra_Diag2CSR_{nullptr},
+      n_{n},
+      nnz_{0}
+  {}
+
+  hiopLinSolverNonSymSparseCUSOLVER::~hiopLinSolverNonSymSparseCUSOLVER()
+  {
+    if(kRowPtr_){
+      delete [] kRowPtr_;
+    }
+    if(jCol_){
+      delete [] jCol_;
+    }
+    if(kVal_){
+      delete [] kVal_;
+    }
+    if(index_covert_CSR2Triplet_){
+      delete [] index_covert_CSR2Triplet_;
+    }
+    if(index_covert_extra_Diag2CSR_){
+      delete [] index_covert_extra_Diag2CSR_;
+    }
+    //KS: make sure we delete the GPU variables
+    cudaFree(dia_);
+    cudaFree(da_);
+    cudaFree(dja_);
+    cudaFree(devr_);
+    cudaFree(devx_);
+    cudaFree(d_work_);
+    cusparseDestroy(handle_); 
+    cusolverSpDestroy(handle_cusolver_);
+    cublasDestroy(handle_cublas_);
+    cusparseDestroyMatDescr(descr_A_);
+
+    cusparseDestroyMatDescr(descr_M_);
+    cusolverSpDestroyCsrluInfoHost(info_lu_);
+    cusolverSpDestroyGluInfo(info_M_);
+
+    klu_free_symbolic(&Symbolic_, &Common_) ;
+    klu_free_numeric(&Numeric_, &Common_) ;
+    if(mia_ != NULL){ 
+      free(mia_);
+    }
+    if(mja_ != NULL){ 
+      free(mja_);
+    }
+  }
+
+  int hiopLinSolverNonSymSparseCUSOLVER::createM(const int n, 
+                                                 const int nnzL, 
+                                                 const int* Lp, 
+                                                 const int* Li,
+                                                 const int nnzU, 
+                                                 const int* Up, 
+                                                 const int* Ui)
+  {
+    int row;
+    for(int i = 0;i<n;++i) {
+      //go through EACH COLUMN OF L first
+      for(int j = Lp[i];j<Lp[i+1];++j) {
```

Comment:
Fixed!

Reply:
`calloc` is used to avoid initializing the array to zero. In this case, using `std::vector` is probably more appropriate solution. I suggest we take care of it at the same time when we optimize the interface.